pkg/istio-agent/sds-agent.go

```diff
@@ -0,0 +1,431 @@
+// environment:
+// - SDS hostPath if node-agent was used
+// - /etc/certs/key if Citadel or other mounted Secrets are used
+// - root cert to use for connecting to XDS server
+// - CA address, with proper defaults and detection
+type AgentConf struct {
+	// Location of JWTPath to connect to CA. If empty, SDS is not possible.
+	// If set SDS will be used - either local or via hostPath.
+	JWTPath string
+
+	// SDSAddress is the address of the SDS server. Starts with unix: for hostpath mount or built-in
+	// May also be a https address.
+	SDSAddress string
+
+	// CertPath is set with the location of the certs, or empty if mounted certs are not present.
+	CertsPath string
+
+	// RequireCerts is set if the agent requires certificates:
+	// - if controlPlaneAuthEnabled is set
+	// - port of discovery server is not 15010 (the plain text default).
+	RequireCerts bool
+
+	// Expected SAN
+	SAN string
+}
+
+// DetectSDS will check if the JWT token required for local SDS is present, and set additional
+// config options for the in-process SDS agent.
+//
+// The JWT token is currently using a pre-defined
+//
+// If node agent and JWT are mounted: it indicates user injected a config using hostPath, and will be used.
+//
+func DetectSDS(discAddr string, tlsRequired bool) *AgentConf {
+	ac := &AgentConf{}
+
+	istiodHost, discPort, err := net.SplitHostPort(discAddr)
+	if err != nil {
+		log.Fatala("Invalid discovery address", discAddr, err)
+	}
+
+	if _, err := os.Stat(JWTPath); err == nil {
+		ac.JWTPath = JWTPath
+	} else {
+		// Can't use in-process SDS.
+		log.Warna("Missing JWT token, can't use in process SDS ", JWTPath, err)
+
+		if discPort == "15012" {
+			log.Fatala("Missing JWT, can't authenticate with control plane. Try using plain text (15010)")
+		}
+		return ac
+	}
+
+	ac.SDSAddress = "unix:" + LocalSDS
+
+	if _, err := os.Stat("/etc/certs/key.pem"); err == nil {
+		ac.CertsPath = "/etc/certs"
+	}
+	if tlsRequired {
+		ac.RequireCerts = true
+	}
+
+	// Istiod uses a fixed, defined port for K8S-signed certificates.
+	if discPort == "15012" {
+		ac.RequireCerts = true
+		// For local debugging - the discoveryAddress is set to localhost, but the cert issued for normal SA.
+		if istiodHost == "localhost" {
+			istiodHost = "istiod.istio-system"
+		}
+		ac.SAN = istiodHost
+	}
+
+	return ac
+}
+
+// Simplified SDS setup. This is called if and only if user has explicitly mounted a K8S JWT token, and is not
+// using a hostPath mounted or external SDS server.
+//
+// 1. External CA: requires authenticating the trusted JWT AND validating the SAN against the JWT.
+//    For example Google CA
+//
+// 2. Indirect, using istiod: using K8S cert.
+//
+// 3. Monitor mode - watching secret in same namespace ( Ingress)
+//
+// 4. TODO: File watching, for backward compat/migration from mounted secrets.
+func StartSDS(conf *AgentConf, isSidecar bool, podNamespace string) (*sds.Server, error) {
+	applyEnvVars()
+
+	gatewaySdsCacheOptions = workloadSdsCacheOptions
```

Comment:
Sorry, what does this mean?

Reply:
No clue - cut and pasted from node agent (I think it means that same options are used when in gateway mode)